sdl-core/src/main/scala/io/smartdatalake/workflow/dataframe/GenericDataFrame.scala

```diff
@@ -0,0 +1,290 @@
+    groupBy(colsInDf.map(col))
+      .agg(Seq(helper.count(col("*")).as(countColname)))
+      .filter(col(countColname) > lit(1))
+  }
+  /**
+   * Returns rows of this data frame which violate uniqueness for specified columns cols.
+   * The result data frame possesses an additional count column countColname.
+   *
+   * @param cols : names of columns which are to be considered, unspecified or empty Array mean all columns of df
+   * @return subdataframe of nlets
+   */
+  def getNonuniqueRows(cols: Seq[String] = schema.columns): GenericDataFrame = {
+    import helper._
+    val dfNonUnique = getNonuniqueStats(cols, "_duplicationCount_").drop("_duplicationCount_")
+    join(dfNonUnique, cols).select(schema.columns.map(col))
+  }
+
+  /**
+   * returns data frame which consists of those rows which violate PK condition for specified columns
+   */
+  def getPKviolators(cols: Seq[String] = schema.columns): GenericDataFrame = getNulls(cols).unionByName(getNonuniqueRows(cols))
+
+  /**
+   * Move partition columns at end of DataFrame as required when writing to Hive in Spark > 2.x
+   */
+  def movePartitionColsLast(partitions: Seq[String])(implicit helper: DataFrameSubFeedCompanion): GenericDataFrame = {
+    import helper._
+    val (partitionCols, nonPartitionCols) = schema.columns.partition(c => partitions.contains(c))
+    val newColOrder = nonPartitionCols ++ partitionCols
+    select(newColOrder.map(col))
+  }
+
+  /**
+   * Convert column names to lower case
+   */
+  def colNamesLowercase(implicit helper: DataFrameSubFeedCompanion): GenericDataFrame = {
+    import helper._
+    select(schema.columns.map(c => col(c).as(c.toLowerCase())))
+  }
+
+  /**
+   * symmetric difference of two data frames: (df∪df2)∖(df∩df2) = (df∖df2)∪(df2∖df)
+   *
+   * @param other         : data frame to compare with
+   * @param diffColName : name of boolean column which indicates whether the row belongs to df
+   * @return data frame
+   */
+  def symmetricDifference(other: GenericDataFrame, diffColName: String = "_in_first_df"): GenericDataFrame = {
+    require(schema.columns.map(_.toLowerCase).toSet == other.schema.columns.map(_.toLowerCase).toSet, "DataFrames must have the same columns for symmetricDifference calculation")
+    // reorder columns according to the original df for calculating symmetricDifference
+    val colOrder = schema.columns.map(helper.col)
+    val dfOtherPrep = other.select(colOrder)
+    this.except(dfOtherPrep).withColumn(diffColName, helper.lit(true))
+      .unionByName(dfOtherPrep.except(this).withColumn(diffColName, helper.lit(false)))
+  }
+
+  /**
+   * compares df with df2
+   *
+   * @param other : data frame to comapre with
+   * @return true if both data frames have the same cardinality, schema and an empty symmetric difference
+   */
+  def isEqual(other: GenericDataFrame): Boolean = {
+    // As a set-theoretic function symmetricDifference ignores multiple occurences of the same row.
```

Comment:
Well, in a set multiple occurences of the same object are ignored.
As symmetricDifference is a set theoretic function, it's implemented like that.
But isEqual should not ignore multiple occurences of the same object are ignored. Thats why there is an additional condition on the count (you can ask @kaelte about details of set theory).
Note that this is also code moved during refactoring, and not new...

Reply:
`symmetricDifference` is based on [DataSet.except](https://spark.apache.org/docs/latest/api/scala/org/apache/spark/sql/Dataset.html#except(other:org.apache.spark.sql.Dataset[T]):org.apache.spark.sql.Dataset[T]) which removes n-lets.

Since Spark 2.4.0 there exists [DataSet.exceptAll](https://spark.apache.org/docs/latest/api/scala/org/apache/spark/sql/Dataset.html#exceptAll(other:org.apache.spark.sql.Dataset[T]):org.apache.spark.sql.Dataset[T]) which _preserves the duplicates_ (even if they occure more than twice).

If the definition of `symmetricDifference` is changed to using [DataSet.exceptAll](https://spark.apache.org/docs/latest/api/scala/org/apache/spark/sql/Dataset.html#exceptAll(other:org.apache.spark.sql.Dataset[T]):org.apache.spark.sql.Dataset[T]) then `isequal` will not need to compare anymore the cardinalities.

In Mathematics, a set contains only one instance of each element, in contrast to a [multiset](https://en.wikipedia.org/wiki/Multiset). A Spark Dataset is actually a Datamultiset.